runtime/compiler/trj9/z/codegen/J9TreeEvaluator.cpp

```diff
@@ -15832,31 +15768,116 @@ TR::Register *
+         break;
+         }
+      case TR::i2pd:
+      case TR::l2pd:
+      case TR::pdshlOverflow:
+      case TR::pdshr:
+         {
+         firstCallParamIndex = 2;
+         isResultPD = true;
+         break;
+         }
+      // Handle variable precision
+      case TR::lcall:
+      case TR::icall:
+         {
+         switch (pdopNode->getSymbol()->getMethodSymbol()->getMethod()->getRecognizedMethod())
+            {
+            case TR::com_ibm_dataaccess_DecimalData_convertPackedDecimalToInteger_:
+            case TR::com_ibm_dataaccess_DecimalData_convertPackedDecimalToInteger_ByteBuffer_:
+            case TR::com_ibm_dataaccess_DecimalData_convertPackedDecimalToLong_:
+            case TR::com_ibm_dataaccess_DecimalData_convertPackedDecimalToLong_ByteBuffer_:
+               {
+               isVariableParam = true;
+               break;
+               }
 
-   // Avoid evaluating an evaluated pdOpNode (first child of BCDCHK) under a BCDCHK node.
-   // This is to avoid generating unnecessary labels that points to non-existant OOL paths
-   // that can eventually cause RA to produce incorrect register use counts.
-   if(rv != NULL)
+            default: TR_ASSERT_FATAL(0, "BCDCHKEvaluator: Could not find recognized method for variable precision DAA operation.\n");
+            }
+         break;
+         }
+
+      default: TR_ASSERT_FATAL(0, "Operation Code not supported by BCDCHKEvaluator.\n");
+      }
+
+   // Evaluate call parameters
+   TR::Node* callParamRoot = isVariableParam ? pdopNode : node;
+   for (uint32_t i = firstCallParamIndex; i < callParamRoot->getNumChildren(); ++i)
+      {
+      if (!callParamRoot->getChild(i)->isSingleRefUnevaluated())
+         cg->evaluate(callParamRoot->getChild(i));
+      }
+
+   /*
+    * Avoid evaluating an evaluated pdOpNode (first child of BCDCHK) under a BCDCHK node if
+    * it is already evaluated.
+    *
+    * This is to avoid generating OOL paths without mainline sequences. OOL without mainline can
+    * cause RA to produce incorrect register use counts, and eventually produce incorrect GC maps that
+    * make GC fail during runtime.
+    */
+   if (resultReg != NULL)
       {
-      for(int i = 0; i < node->getNumChildren(); ++i)
+      if (isVariableParam)
+         cg->recursivelyDecReferenceCount(pdopNode);          // variable parameter l2pd is a call node
+      else
          {
-         cg->recursivelyDecReferenceCount(node->getChild(i));
+         // first child
+         cg->decReferenceCount(pdopNode);
+
+         // second child
+         if (isResultPD)
+            cg->recursivelyDecReferenceCount(node->getSecondChild());
+
+         // call parameters: 2nd/3rd and above
+         for (uint32_t i = firstCallParamIndex; i < node->getNumChildren(); ++i)
+            cg->decReferenceCount(node->getChild(i));
          }
-      return rv;
+
+      traceMsg(cg->comp(), "Skipping BCDCHK node n%dn\n", node->getGlobalIndex());
+      return resultReg;
```

Comment:
Perhaps using an `else` path may be more clear here rather than having multiple return points. It gets hard to reason about the lifetime of a function this way.